File: tools/get-includes.pl
#!/usr/bin/env perl

use strict;
use warnings;
use File::Basename;
use File::Spec;

my $THIS_FILE = File::Spec->rel2abs (File::Spec->curdir (), __FILE__);
my $THIS_DIR = dirname ($THIS_FILE);

sub usage ($)
{
  my ($exit_code) = @_;
  print <<EOF;
Usage: $THIS_FILE filename ...
EOF

  exit ($exit_code) if $exit_code != 0;
}

sub print_includes (@)
{
  my (@filenames) = @_;

  foreach (@filenames)
    {
      open (my $fh, "<", "$_") or die "Failed to open $_: $!\n";

      my $fmt_name = basename($_);
      my @local_includes = ();
      my @system_includes = ();

      while (<$fh>)
        {
          chomp;
          if ($_ =~ /\s*#\s*include\s*\"(.+?)\"/)
            {
              push (@local_includes, $1);
            }
          elsif ($_ =~ /\s*#\s*include\s*<(.+?)>/)
            {
              push (@system_includes, $1)
            }
        }

      my $local_count = scalar (@local_includes);
      my $system_count = scalar (@system_includes);

      if ($local_count > 0 || $system_count > 0)
        {
          print ("File: $fmt_name\n");

          if ($local_count > 0)
            {
              print ("Local includes:\n");
              print (join ("\n", @local_includes) . "\n\n");
            }

          if ($system_count > 0)
            {
              print ("System includes:\n");
              print (join ("\n", @system_includes) . "\n\n");
            }
        }

      close ($fh);
    }
}

if (scalar (@ARGV) < 1)
  {
    usage (1);
  }

print_includes (@ARGV);

File: tools/get-includes.py
#!/usr/bin/env python3

import os
import sys
import re
from pathlib import Path, PurePath

THIS_FILE = Path(__file__).resolve()
THIS_DIR = THIS_FILE.parent

# Regular expression to match a local include "#include "filename.h""
# Follows traditional preprocessor syntax where no spaces leading the #.
local_include_regexp = re.compile(r"^#\s*?include\s+?\"(.+?)\"", re.MULTILINE)

# Regular expression to match a system include "#include <filename.h>"
# Follows traditional preprocessor syntax where no spaces leading the #.
system_include_regexp = re.compile(r"^#\s*?include\s+?<(.+?)>", re.MULTILINE)

def read_file(filename: Path) -> str:
    lines = []
    try:
        with open(filename, "r", encoding="utf-8") as file:
            for line in file:
                lines.append(line)
    finally:
        return "".join(lines)

def main() -> None:
    argv = sys.argv
    argc = len(argv)

    if argc < 2:
        print(f"{THIS_FILE} files ...")
        exit(1)

    files = [Path(x).resolve() for x in argv[1:]]

    for file in files:
        content = read_file(file)
        local_includes = re.findall(local_include_regexp, content)
        system_includes = re.findall(system_include_regexp, content)
        if not local_includes and not system_includes:
            continue
        print(f"Filename: {file.name}")
        if system_includes:
            system_includes = "\n".join(system_includes)
            print(f"System includes:")
            print(f"{system_includes}")
        if local_includes:
            local_includes = "\n".join(local_includes)
            print(f"Local includes:")
            print(f"{local_includes}")
        print("")

if __name__ == "__main__":
    main()
